numpy/core/src/umath/ufunc_object.c

```diff
@@ -4456,24 +4487,43 @@ PyUFunc_GenericReduction(PyUFuncObject *ufunc, PyObject *args,
     }
     else if (operation == UFUNC_ACCUMULATE) {
         if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|OO&O&:accumulate",
-                                        accumulate_kwlist,
-                                        &op,
-                                        &axes_in,
-                                        PyArray_DescrConverter2, &otype,
-                                        PyArray_OutputConverter, &out)) {
+                                         accumulate_kwlist,
+                                         &op,
+                                         &axes_in,
+                                         PyArray_DescrConverter2, &otype,
+                                         PyArray_OutputConverter, &out)) {
             goto fail;
         }
     }
     else {
-        if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|OO&O&iO:reduce",
+        if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|OO&O&iOO:reduce",
                                         reduce_kwlist,
                                         &op,
                                         &axes_in,
                                         PyArray_DescrConverter2, &otype,
                                         PyArray_OutputConverter, &out,
-                                        &keepdims, &initial)) {
+                                        &keepdims, &initial, &where)) {
             goto fail;
         }
+        /* Interpret mask */
+        if (where != NULL) {
+            PyArray_Descr *dtype;
+            dtype = PyArray_DescrFromType(NPY_BOOL);
+            if (dtype == NULL) {
+                goto fail;
+            }
+            /*
+             * Optimization: where=True is the same as no where argument.
+             * This lets us document it as a default argument.
+             */
```

Comment:
This optimization needs to come before calling `PyArray_DescrFromType`, else you leak a `dtype`

Reply:
Actually, I think you might leak it no matter what.

Reply:
`PyArray_FromAny` steals the reference to the `dtype`, so indeed the `dtype` should be created inside the `if` clause.

Reply:
Actually, turns out I wrote a `_wheremask_converter` when redoing the keyword parsing for the normal `ufunc`. Might as well use it...